lib/ABI/FunctionType/Conversion.cpp

```diff
@@ -45,79 +47,99 @@ class ToCABIConverter {
 
 private:
   const abi::Definition &ABI;
+  model::Binary &Binary;
+  TypeBucket Bucket;
 
 public:
-  explicit ToCABIConverter(const abi::Definition &ABI) : ABI(ABI) {
-    ABI.verify();
-  }
-
-public:
-  /// Entry point for the `toCABI` conversion.
-  std::optional<model::TypePath>
-  tryConvert(const model::RawFunctionType &Function,
-             TupleTree<model::Binary> &Binary) const;
-
-private:
-  template<bool DryRun = false>
+  ToCABIConverter(const abi::Definition &ABI, model::Binary &Binary) :
+    ABI(ABI), Binary(Binary), Bucket(Binary) {}
+  ~ToCABIConverter() { Bucket.drop(); }
```

Comment:
I can expand it in the comment, if you want. That's intentional. The intentions are as follows:
```cpp
void whatever() {
  TypeBucket Bucket = getFromSomewhere();
  
  // do stuff
  
  // `Bucket` automatically commits its contents unless it was explicitly dropped (in which case it because empty, so there's nothing to commit).
}
```

```cpp
std::optional<model::CABIFunctionType> whatever() {
  ToCABIConverter ToCABI(/*whatever*/);
  
  ToCABI.Bucket = insertSomeTypesFromSomewhere();
  
  if (someRandomCheck()) {
    // Check failed, we have to abort.
    // Luckily, the converter properly cleans up our bucket for us, so we don't need to do anything extra.
    return std::nullopt;
  }

  // do stuff
  
  ToCABI.Bucket.commit(); // Commit the bucket explicitly to ensure the original `Binary` is not written into unnecessarily.
  return Result;
}
```

Reply:
I feel like in this specific case RAII is more confusing that helpful.
Can we force the develop to be explicit and call `discard` or `commit`? Also, we can `assert` in the destructor than one of the twos has been called.

It's not obvious from the name that things are going to happen upon destruction.

Reply:
I guess I'm fine with dropping the RAII on the `Converter` - I guess `ToCABI.drop()`s would make sense in the failing branches.

But I'd say it makes sense to keep the `TypeBucket` behaviour, i.e. auto-commit unless manually dropped. Are we on the same page?

P. S. Love the assert idea.

Reply:
Once we inline primitives `TypeBucket` will be able to behave like an actual bucket and just take a `const &` to `Binary`, it'd make much more sense.

Then we can have a `commit` method that takes a non-`const` reference.

Autocommiting magic is superfluous magic.

Reply:
I guess that approach would also make a lot of sense. I'm sold.

Reply:
Done: https://github.com/revng/revng/pull/232/commits/0d3ebdc2dd8e6191cbb866bba6b091a3a2b454b9.